_data/build-research.py

```diff
@@ -1,85 +1,209 @@
-# a script to automatically generate citations for papers the research list component
-
-# before running:
-# install python                    https://www.python.org/downloads/
-# install pip                       https://pip.pypa.io/en/stable/installing/
-# install manubot                   https://github.com/manubot/manubot#installation
-
+# packages
 import sys
 import os
 import json
 import yaml
 import subprocess
 from datetime import datetime
+from yaml.loader import SafeLoader
+
+# allow printing ANSI color codes
+os.system("")
+
+# yaml loader with line numbering
+# https://stackoverflow.com/questions/13319067/parsing-yaml-return-with-line-number
+class SafeLineLoader(SafeLoader):
+    def construct_mapping(self, node, deep=False):
+        mapping = super(SafeLineLoader, self).construct_mapping(node, deep=deep)
+        if node.start_mark.column == 2:
+            mapping["__line_number__"] = node.start_mark.line + 1
+        return mapping
 
 # input and output files
 current_dir = os.path.dirname(os.path.realpath(__file__))
 input_file = os.path.join(current_dir, "research-input.yml")
 output_file = os.path.join(current_dir, "research-output.yml")
 
-# load input papers as yaml
+# log wih color
+def log(message, color="white"):
+    palette = {
+        "white": "\033[97m",
+        "red": "\033[91m",
+        "blue": "\033[94m",
+        "green": "\033[92m",
+        "yellow": "\033[93m",
+        "reset": "\033[0m"
+    }
+    print(f"{palette[color]}{message}{palette['reset']}")
+
+# log and exit
+def exit(message):
+    log(message, "red")
+    sys.exit(1)
+
+# check that input file exists
+if not os.path.isfile(input_file):
+    exit("Can't find research-input.yml")
+
+# try to open input file
 try:
     with open(input_file, encoding="utf8") as file:
-        input_papers = yaml.load(file, Loader=yaml.FullLoader)
+        # try to parse input file as yaml
+        try:
+            input_papers = yaml.load(file, Loader=SafeLineLoader)
+        except Exception:
+            exit("Can't parse research-input.yml. Make sure it's valid YAML.")
 except Exception:
-    print("Problem with input file.")
-    sys.exit(1)
+    exit("Can't open research-input.yml")
 
-# load existing output papers as yaml
+# check that top level of input yaml is list/array
+if type(input_papers) != list:
+    exit("research-input.yml not in expected format")
+
+# try to open existing output file and parse as yaml
 try:
     with open(output_file, encoding="utf8") as file:
-        output_papers = yaml.load(file, Loader=yaml.FullLoader)
+        output_papers = yaml.load(file, Loader=SafeLineLoader)
+        if type(output_papers) != list:
+            raise
```

Comment:
`raise` is not in an exception handling context, so no exception will be re-raised here. You probably meant:
```
raise Exception()
```